### PR TITLE
Add source local scaling Kafka Source Connectors [HZ-1807]

### PR DESCRIPTION
**Pull request by ldziedziul**
Adds scaling support for Kafka Connect sources using local parallelism

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [x] Send backports/forwardports if fix needs to be applied to past/future releases
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ConnectorWrapper.java [hunk 1]
```diff
@@ -70,32 +76,24 @@ private static SourceConnector newConnectorInstance(String connectorClazz) {
...
-            taskRunner.stop();
-        } finally {
-            connector.stop();
-        }
+    private SourceTask createSourceTask() {
+        Class<? extends SourceTask> taskClass = connector.taskClass().asSubclass(SourceTask.class);
+        return newInstance(Thread.currentThread().getContextClassLoader(), taskClass.getName());
     }
 
     public Map<Map<String, ?>, Map<String, ?>> createSnapshot() {
```
Comment (orcunc) [inline extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ConnectorWrapper.java:99]: createSnapshot is not used

File: hazelcast/src/main/java/com/hazelcast/jet/core/ProcessorMetaSupplier.java [hunk 2]
```diff
@@ -484,7 +505,7 @@ class SpecificMemberPms implements ProcessorMetaSupplier, IdentifiedDataSerializ
         SpecificMemberPms() {
         }
 
-        private SpecificMemberPms(ProcessorSupplier supplier, Address memberAddress) {
+        private SpecificMemberPms(ProcessorSupplier supplier, Address memberAddress, int localParallelism) {
```
Comment (orcunc) [inline hazelcast/src/main/java/com/hazelcast/jet/core/ProcessorMetaSupplier.java:508]: localParallelism is not used

**Review by frant-hartm — CHANGES_REQUESTED** [review]

File: hazelcast/src/main/java/com/hazelcast/jet/core/ProcessorMetaSupplier.java (hunk 2 above)
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/jet/core/ProcessorMetaSupplier.java]: thanks, fixed https://github.com/hazelcast/hazelcast/pull/23831/commits/d1f93791d7890a65796e849c7853a9938d35b8c3

File: extensions/kafka-connect/src/test/java/com/hazelcast/jet/kafka/connect/impl/KafkaConnectSourceTest.java [hunk 3]
```diff
@@ -1,163 +0,0 @@
...
-import java.util.stream.Collectors;
-
-import static com.hazelcast.jet.kafka.connect.impl.DummySourceConnector.INSTANCE;
-import static com.hazelcast.jet.kafka.connect.impl.DummySourceConnector.ITEMS_SIZE;
-import static java.util.Collections.singletonList;
-import static org.assertj.core.api.Assertions.assertThat;
-import static org.assertj.core.api.Assertions.assertThatThrownBy;
-
-@Category({QuickTest.class, ParallelJVMTest.class})
-public class KafkaConnectSourceTest {
```
[the remark this answers is not in the document]
Comment (ldziedziul) [reply extensions/kafka-connect/src/test/java/com/hazelcast/jet/kafka/connect/impl/KafkaConnectSourceTest.java:45]: Added here: https://github.com/hazelcast/hazelcast/pull/23831/commits/7083ace122d7435f8593b3d7816498a68b750ac8

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java [hunk 4]
```diff
@@ -71,13 +79,33 @@ public static StreamSource<SourceRecord> connect(@Nonnull Properties properties)
...
         checkRequiredProperty(properties, "connector.class");
 
-        return SourceBuilder.timestampedStream(name, ctx ->
-                        new KafkaConnectSource(properties))
-                .fillBufferFn(KafkaConnectSource::fillBuffer)
-                .createSnapshotFn(KafkaConnectSource::createSnapshot)
-                .restoreSnapshotFn(KafkaConnectSource::restoreSnapshot)
-                .destroyFn(KafkaConnectSource::destroy)
-                .build();
+        int tasksMax = Integer.parseInt(properties.getProperty("tasks.max", "1"));
```
[the remark this answers is not in the document]
Comment (ldziedziul) [reply extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java:82]: Is this enough? https://github.com/hazelcast/hazelcast/pull/23831/commits/158f39023827c2f7f2705c771d485677af08d379

File: hazelcast/src/main/java/com/hazelcast/jet/core/ProcessorMetaSupplier.java [hunk 5]
```diff
@@ -467,7 +470,25 @@ static ProcessorMetaSupplier forceTotalParallelismOne(
...
+    }
+
+
+    /**
+     * Wraps the provided {@code ProcessorSupplier} into a meta-supplier that
+     * will only use the given {@code ProcessorSupplier} on a random node with the
+     * given localParallelism.
+     *
+     * @param supplier                  the supplier that will be wrapped
+     * @param preferredLocalParallelism the value to return from {@link #preferredLocalParallelism()}
```
[the remark this answers is not in the document]
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/jet/core/ProcessorMetaSupplier.java:484]: Added https://github.com/hazelcast/hazelcast/pull/23831/commits/51fcd571970d043c41c00214bf770a0818f8c208

**Review by orcunc — APPROVED** [review]

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java:82]: Maybe add that `.setLocalParallelism` does not work due to this behaviour (or make it that parallelism = min(local parallelism, task.max))

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        return false;
+    }
+
+    @Override
+    public boolean complete() {
+        if (snapshotInProgress) {
+            return false;
+        }
+        List<SourceRecord> sourceRecords = taskRunner.poll();
+        if (traverser == null) {
```
Comment (TomaszGaweda) [inline extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java:71]: You should poll next records only if traverser is null (= previous one was exhausted)

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            this.traverser = traverser(sourceRecords)
```
Comment (TomaszGaweda) [inline extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java:72]:
```suggestion
            this.traverser = traverser(sourceRecords)
                                      .map(r -> {
                                           taskRunner.commitRecord(r);
                                           return r;
                                      }); 
```

Alternatively you can do `taskRunner.commitRecord` in flatMap inside `traverser` method

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                    .onFirstNull(() -> traverser = null);
+        }
+        emitFromTraverser(traverser);
```
Comment (TomaszGaweda) [inline extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java:75]:
```suggestion
        return emitFromTraverser(traverser);
```

File: extensions/kafka-connect/src/test/java/com/hazelcast/jet/kafka/connect/KafkaConnectDatagenIntegrationTest.java [hunk 9]
```diff
@@ -87,5 +93,98 @@ public void testReadFromDatagenConnector() throws Exception {
         }
     }
 
+    @Test
+    public void testScalingWithDatagenConnector() throws Exception {
```
Comment (TomaszGaweda) [inline extensions/kafka-connect/src/test/java/com/hazelcast/jet/kafka/connect/KafkaConnectDatagenIntegrationTest.java:97]: datagen is already in the class name

File: extensions/kafka-connect/src/test/java/com/hazelcast/jet/kafka/connect/KafkaConnectDatagenIntegrationTest.java [hunk 10]
```diff
@@ -87,5 +93,98 @@ public void testReadFromDatagenConnector() throws Exception {
...
+
+        waitForFirstSnapshot(jobRepository, job.getId(), 30, false);
+
+        hazelcastInstances[1].shutdown();
+
+        assertTrueEventually(() -> assertEquals(2, hazelcastInstances[0].getCluster().getMembers().size()));
+
+        try {
+            job.join();
+            fail("Job should have completed with an AssertionCompletedException, but completed normally");
```
Comment (TomaszGaweda) [inline extensions/kafka-connect/src/test/java/com/hazelcast/jet/kafka/connect/KafkaConnectDatagenIntegrationTest.java:182]: what if processing of incoming data end before job is restored? we do not check if job is still running after shutdown of the member

**Review by TomaszGaweda — CHANGES_REQUESTED** [review]

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (frant-hartm) [reply extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java:82]: That's a good point, maybe we can just take the `localParallelism` instead.

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java [hunk 11]
```diff
@@ -71,13 +79,33 @@ public static StreamSource<SourceRecord> connect(@Nonnull Properties properties)
...
+
+        return Sources.streamFromProcessorWithWatermarks(name, true,
+                eventTimePolicy -> ProcessorMetaSupplier.preferLocalParallelismOnSingleMember(
+                        new ProcessorSupplier() {
+                            private ConnectorWrapper connectorWrapper;
```
Comment (frant-hartm) [inline extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java:87]: not necessary, but adding `transient` makes it clear that it is not serialized

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java [hunk 12]
```diff
@@ -71,13 +79,33 @@ public static StreamSource<SourceRecord> connect(@Nonnull Properties properties)
...
+
+                            @Override
+                            public void init(@Nonnull Context context) {
+                                this.connectorWrapper = new ConnectorWrapper(properties);
+                            }
+
+                            @Override
+                            public void close(@Nullable Throwable error) {
+                                connectorWrapper.stop();
```
Comment (frant-hartm) [inline extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java:96]: connectorWrapper may be null, see javadoc for close

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    private Traverser<Entry<BroadcastKey<String>, TaskRunner.State>> snapshotTraverser;
+    private boolean snapshotsEnabled;
+    private int processorIndex;
+    private Traverser<?> traverser;
+
+    public ReadKafkaConnectP(ConnectorWrapper connectorWrapper, EventTimePolicy<? super SourceRecord> eventTimePolicy) {
+        this.connectorWrapper = connectorWrapper;
+        this.eventTimeMapper = new EventTimeMapper<>(eventTimePolicy);
+        eventTimeMapper.addPartitions(1);
+
```
Comment (frant-hartm) [inline extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java:50]:
```suggestion
```

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/TaskRunner.java [hunk 14]
```diff
@@ -132,7 +156,102 @@ public Map<String, String> configs() {
...
+         * Key represents the partition which the record originated from. Value
+         * represents the offset within that partition. Kafka Connect represents
+         * the partition and offset as arbitrary values so that is why it is
+         * stored as map.
+         * See {@link SourceRecord} for more information regarding the format.
+         */
+        private final Map<Map<String, ?>, Map<String, ?>> partitionsToOffset;
+
+        State() {
+            this(new ConcurrentHashMap<>());
```
Comment (frant-hartm) [inline extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/TaskRunner.java:213]: We can assign it directly to the field here.

**Review by frant-hartm — CHANGES_REQUESTED** [review]

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java (hunk 8 above)
Comment (ldziedziul) [reply extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java:75]: Shouldn't we return always `false`?

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java (hunk 6 above)
Comment (ldziedziul) [reply extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java:69]: fixed https://github.com/hazelcast/hazelcast/pull/23831/commits/711671ab9ba30b3670bd400acd034838182fad0c

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ldziedziul) [reply extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java]: fixed: https://github.com/hazelcast/hazelcast/pull/23831/commits/f830383d676b683ca14101a40abe02781a28820d

File: extensions/kafka-connect/src/test/java/com/hazelcast/jet/kafka/connect/KafkaConnectDatagenIntegrationTest.java (hunk 9 above)
Comment (ldziedziul) [reply extensions/kafka-connect/src/test/java/com/hazelcast/jet/kafka/connect/KafkaConnectDatagenIntegrationTest.java]: renamed: https://github.com/hazelcast/hazelcast/pull/23831/commits/b10441acecac5e8713d08f1c0c625f31effa7d09

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java (hunk 11 above)
Comment (ldziedziul) [reply extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java]: fixed: https://github.com/hazelcast/hazelcast/pull/23831/commits/dc7f65e44e6beae45281092eba18cc60a51ac754

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java (hunk 12 above)
Comment (ldziedziul) [reply extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/KafkaConnectSources.java]: fixed: https://github.com/hazelcast/hazelcast/pull/23831/commits/18b1798181ec279cda16e2801225428b3f0806fa

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java (hunk 13 above)
Comment (ldziedziul) [reply extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/ReadKafkaConnectP.java]: fixed https://github.com/hazelcast/hazelcast/pull/23831/commits/aeb4d6d7ad5524585d9cd36598eef62685f7f54f

File: extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/TaskRunner.java (hunk 14 above)
Comment (ldziedziul) [reply extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/TaskRunner.java]: fixed https://github.com/hazelcast/hazelcast/pull/23831/commits/5839ab6a3b00631f55381d5fbbb2e305dea5eb13

File: extensions/kafka-connect/src/test/java/com/hazelcast/jet/kafka/connect/KafkaConnectDatagenIntegrationTest.java (hunk 10 above)
Comment (ldziedziul) [reply extensions/kafka-connect/src/test/java/com/hazelcast/jet/kafka/connect/KafkaConnectDatagenIntegrationTest.java]: added assertion https://github.com/hazelcast/hazelcast/pull/23831/commits/2d5e907883c5692205d7c168a69e68b839cfd728

**Review by TomaszGaweda — APPROVED** [review]
:party-parrot:

**Pull request merged**